File: 백준/Gold/1647. 도시 분할 계획/도시 분할 계획.cc
// Authored by : chjh2129
#include <bits/stdc++.h>

using namespace std;

using ti = tuple<int,int,int>;

int n, m;
vector<ti> edge;
vector<int> p(100005, -1);

int find(int x){ return (p[x] < 0 ? x : p[x] = find(p[x])); }

bool merge(int x, int y){
  x = find(x);
  y = find(y);
  if(x == y) return 0;
  if(p[x] == p[y]) p[x]--;
  if(p[x] > p[y]) swap(x, y);
  p[y] = x;
  return 1;
}

int main(){
  cin.tie(nullptr)->sync_with_stdio(false);

  // input
  cin >> n >> m;
  while(m--){
    int u, v, c;
    cin >> u >> v >> c;
    edge.push_back({c, u, v});
  }

  // solve
  sort(edge.begin(), edge.end());
  int town = n; // 현재 분리된 마을의 개수
  int ans{}; // 유지비의 최솟값
  for(auto &[cost, u, v] : edge){
    if(town == 2) break; // 분리된 마을이 2개면 패스
    if(!merge(u, v)) continue;
    town--;
    ans += cost;
  }

  // output
  cout << ans;
}

File: 백준/Gold/1647. 도시 분할 계획/도시 분할 계획.py
#Authored by : chjh2129
import sys

input = sys.stdin.readline

def find(x: int, p: list[int]) -> int:
  root: int = x
  while p[root] > 0:
    root = p[root]
  
  while p[x] > 0:
    nxt, p[x] = p[x], root
    x = nxt

  return root

def union(x: int, y: int, p: list[int]) -> bool:
  x = find(x, p)
  y = find(y, p)
  if x == y: return False
  if p[x] == p[y]: p[x] -= 1
  if p[x] > p[y]: x, y = y, x
  p[y] = x
  return True

def main() -> None:
  # input & init
  N, M = map(int, input().split())
  Edge: list[tuple] = [ (u, v, cost) for u, v, cost in (map(int, input().split()) for _ in range(M)) ]
  P: list[int] = [ -1 for _ in range(N+1) ]

  # solve
  Edge.sort(key = lambda x: x[2])

  '''
    마을을 총 2개로 나누어야 하므로 cnt가 2가 되면 종료
  '''
  cnt, ans = N, 0
  for u, v, cost in Edge:
    if cnt == 2: break
    if union(u, v, P):
      cnt -= 1
      ans += cost

  # output
  print(ans)

sys.exit(main())